# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [6]:
def add_two(x):
    return x + 2

print(add_two(4))

#lambda version...
# without using variable to pass in data

print((lambda x: x+2)(4))

# using a variable
num = 4
print((lambda x: x+2)(num))

6
6
6


#### Saving to a Variable

In [7]:
f_test = lambda x: x + 2
f_test(4)

6

#### Multiple Inputs

In [13]:
# multiple inputs with no variable

print((lambda x, y, z: x * y * z)(3, 5, 8))

# Multiple inputs with a variable attached
x = lambda x, y, z: x * y * z

print(x(3, 5, 8))

120
120


#### Passing a Lambda into a Function

In [ ]:
def multiply(f, num):
    '''
        f expects a lambda function
        num expects a number
    '''
    return f(num)

multiply(lambda x: x * x, 4)

#### Returning a Lambda from a Function

In [18]:
# Regular defined function
def mutliply_test(num):
    return num * 4

#function within a function:
def return_func():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = return_func()
print(return_func())
print(f_return(4))

# lambda function returned from a regular function
def return_lam(b, c):
    return lambda x, a: x + a + b + c

r_lamb = return_lam(4,5)
print(r_lamb(5,5))

<function return_func.<locals>.multiply at 0x10401fa60>
8
<function return_lam.<locals>.<lambda> at 0x10401fe20>
19


#### If Statements within Lambdas

In [20]:
f_condition = lambda num : num*2 if num >10 else num + 2

# lambda takes one argument, if num > 10: return num * 2, else: return num + 2

print(f_condition(16))

32


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [23]:
f  = lambda x : x ** 3

print(f(3))
cubed = f(3)
print(cubed)

27
27


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [26]:
# map(func, iterable)

def squared(num1, num2):
    if num1 < 10 and num2 < 10:
        return num1 ** 2, num2 ** 2
    else:
        return num1, num2

numbers = [4, 11, 20, 3, 15, 20]
more_nums = [4, 10, 3, 2, 6]

squared_nums_map = list(map(squared, numbers, more_nums))
print(squared_nums_map)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [35]:
lambda_map = list(map(lambda x, y: (x**2, y**2) if x < 10 and y < 10 else (x, y), numbers, more_nums))
print(lambda_map)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [36]:
numbies = [x for x in range(10)]
print(list(map(lambda x : (x*2) - 1, numbies)))

[-1, 1, 3, 5, 7, 9, 11, 13, 15, 17]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [39]:
names = ['Bob', 'Andy', 'Max', 'Evan', 'Ami', 'Angela', 'Erin']

def a_names(names):
    if names[0].lower() == 'a':
        return True
    else:
        return False

new_names_list = list(filter(a_names, names))

print(new_names_list)

# Filter always makes use of a function that returns a Bool value

['Andy', 'Ami', 'Angela']


#### Using Lambda's with Filter()

In [41]:
new_names_lambda = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lambda)

['Andy', 'Ami', 'Angela']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [55]:
from statistics import mean
import random

many_numbers = [random.randint(1, 10) for x in range(50)]
many_mean = mean(many_numbers)
less_than_mean = list(filter(lambda x : True if x < many_mean else False, many_numbers))
print(many_mean)
print(less_than_mean)
print(len(less_than_mean))

5.56
[2, 1, 1, 3, 4, 2, 5, 4, 4, 2, 3, 3, 2, 4, 1, 3, 5, 1, 5, 1, 2, 5, 3, 4, 5, 5]
26


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [58]:
from functools import reduce

#reduce(function, iterable)

list1 = [2, 4 , 6, 8, 10]

def add_nums(num1, num2):
    return num1 + num2

result_add = reduce(add_nums, list1)
print(result_add)

def subtract_nums(num1, num2):
    return num1 - num2

result_sub = reduce(subtract_nums, list1)
print(result_sub)

30
-26


#### Using Lambda's with Reduce()

In [59]:
result_lamb = reduce(lambda x, y: x + y, list1)
print(result_lamb)


30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [60]:
my_list = [1, 2, 3, 4]

result = reduce(lambda x, y: x * y, my_list)
print(result)

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [81]:
def add_nums(num):
    # Here is our base case...
    if num <= 1:
        print('add_nums(1) = 1')
    else:
        print(f'add_nums({num}) = {num} + add_nums({num - 1})')
        return num + add_nums(num-1)

add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


TypeError: unsupported operand type(s) for +: 'int' and 'NoneType'

#### Writing a Factorial Function

In [87]:
def num_factorial(num):
    if num == 1:
        return 1
    else:
        return num * num_factorial(num-1)

    
print(num_factorial(10))

3628800


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [90]:
def sub_all(num, *args):
    if not args:
        return
    else:
        return num - sub_all(args[0], args[1:])
    
def sub_range(num):
    if num <= 1:
        return 1
    else:
        return num - sub_range(num-1)

print(sub_range(10))

5


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [92]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start # yield keyword denotes a generator
        start += step
        
for i in my_range(20, 0):
    print(i)
    
my_range(20, 2)

0
2
4
6
8
10
12
14
16
18


<generator object my_range at 0x1059382b0>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

# removing the while loop above would create an infinite generator

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [98]:
def decreasing_squares(num):
    while num > 0:
        yield num**2
        num -= 1

for i in decreasing_squares(5):
    print(i)

25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [75]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

only_places = list(filter(lambda word: False if word.strip(' ') == '' else True, places))
print(only_places)



['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [79]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key = (lambda name : name.split(' ')[-1].lower()))
print(author)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [157]:
# F = (9/5)*C + 32

places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

city, temp_c = zip(*places)
converted = list(zip(city, map(lambda x : (9/5) * x + 32, temp_c)))
print(converted)




[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [164]:
def fib_recursive(num):
    if num == 0:
        return 0
    elif num == 1 or num == 2:
        return 1
    else:
        return fib_recursive(num - 1) + fib_recursive(num - 2)
    
def fib(num):
    for i in range(num+1):
        print(f'Iteration {i}: {fib_recursive(i+1)}')

fib(5)

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
